code/io/spacemouse.cpp

```diff
@@ -0,0 +1,179 @@
+#include "spacemouse.h"
+
+#include <hidapi.h>
+
+using namespace io::spacemouse;
+
+#pragma pack(push,1)
+struct connexion_3d_old_protocol {
+	uint8_t type;
+	union {
+		struct {
+			int16_t trans_x, trans_y, trans_z;
+		} type1;
+		struct {
+			int16_t rot_x, rot_y, rot_z;
+		} type2;
+		struct {
+			uint8_t buttons[4];
+		} type3;
+	} data;
+};
+
+struct connexion_3d_new_protocol {
+	uint8_t type;
+	union {
+		struct {
+			int16_t trans_x, trans_y, trans_z;
+			int16_t rot_x, rot_y, rot_z;
+		} type1;
+		struct {
+			uint8_t buttons[4];
+		} type3;
+	} data;
+};
+#pragma pack(pop)
+
+void SpaceMouse::poll() {
+	int bytes_read = 0;
+	do {
+		//Read until we have all reports, since we need the last one in the buffer for each type as the most up to date one.
+
+		switch (m_definition.protocol) {
+		case SpaceMouseDefinition::Protocol::CONNEXION_3D_OLD: {
+			static connexion_3d_old_protocol data;
+
+			bytes_read = hid_read(m_deviceHandle, reinterpret_cast<unsigned char*>(&data), sizeof(connexion_3d_old_protocol));
+			if (bytes_read <= 0)
+				continue;
+
+			switch (data.type) {
+			case 1:
+				m_current.translation.xyz.x = static_cast<float>(data.data.type1.trans_x) / 150.0f;
+				m_current.translation.xyz.y = static_cast<float>(data.data.type1.trans_y) / -150.0f;
+				m_current.translation.xyz.z = static_cast<float>(data.data.type1.trans_z) / -150.0f;
+				break;
+			case 2:
+				m_current.rotation.p = static_cast<float>(data.data.type2.rot_x) / -350.0f;
+				m_current.rotation.b = static_cast<float>(data.data.type2.rot_y) / 350.0f;
+				m_current.rotation.h = static_cast<float>(data.data.type2.rot_z) / 350.0f;
```

Comment:
Where do these divisors come from?

Reply:
So, the sign of the divisiors is set so that the reported value is in the right direction for FSO's coordinate system.
The value is kind of arbitrary, really. I started with 350 as that's what other FOSS implementations use (presumably to get the 2^15 down to roughly -100 to 100, I guess?), but then translation was much too slow in FRED, even considering the speed settings. So I just bumped it a bit.
If someone wants to change that, the nonlinearity settings should suffice just fine for that.

Reply:
All right